### ISSUE TITLE
Execute initialize*StagingRepository after sign*Publication

### ISSUE THREAD

**Issue by szpak** [gradle-nexus/publish-plugin#16]
As signing might fail for variety of reasons, I wonder, if it wouldn't be worth to tune the task orchestration to perform it (signing) before (noticeable long and generating side effects - a new staging repository in Nexus) `initialize*StagingRepository` task(s)?

```
$ gw clean publishToSonatype --console=verbose

> Configure project :
> Task :clean
> Task :compileJava
> Task :processResources NO-SOURCE
> Task :classes
> Task :jar
> Task :javadoc
> Task :javadocJar
> Task :sourcesJar
> Task :generateMetadataFileForMavenJavaPublication
> Task :generatePomFileForMavenJavaPublication
> Task :initializeSonatypeStagingRepository      <--- here
> Task :signMavenJavaPublication
> Task :publishMavenJavaPublicationToSonatypeRepository
> Task :publishToSonatype
```

**Comment by marcphilipp** [comment]
Makes sense! Should be fixed by adding `shouldRunAfter(tasks.withType<Sign>())`.

**Comment by vlsi** [comment]
I don't think it should be the default behavior.

I guess sign task fails less often than `compileJava`

**Comment by vlsi** [comment]
> long and generating side effects

Starting `initialize*StagingRepository` enables us to have concurrency, so the repository is ready by the time artifacts are ready for deployment.

**Comment by szpak** [comment]
For fully the configured CI environment signing should be reliable. However, for manual releasing there could be some issues (especially if you a passphrase kept protected and needs to be manually unlocked).

In my projects usually tests takes the biggest amount of time. `sign*` tasks don't depend on them and there is usually nothing between `sign*` and `initialize*StagingRepository` (which with Sonatype Nexus takes even 20+ seconds). Therefore, I don't see a visible negative impact on that switch. However, maybe in your projects/configuration @visi it would be more problematic?

**Comment by vlsi** [comment]
`initialize*StagingRepository` can fail as well. If it fails, then the other tasks make little sense.

> In my projects usually tests takes the biggest amount of time

I do not execute `tests` as a part of `release`.
It happens the other way around:
1) The release is prepared and published for review. The staging repository is closed so everybody knows the release is immutable
2) Everybody validates the release (e.g. run tests on their environments)
3) If everybody's happy, then the release is promoted

There's no sense in running tests as a part of `release`.

**Comment by szpak** [comment]
> There's no sense in running tests as a part of release.

I many (smaller and/or more agile) projects if the (automatic functional/acceptance) tests pass the binaries are ready to be released. The Apache Foundation has different workflow and it might be good for it/them. Nevertheless, I'm not sure that how many non-Apache projects do the same.

Do you think @visi that the change would significantly impact the total time of releasing in your case(s)?

**Comment by vlsi** [comment]
I use the same approach for non-Apache projects as well.
In those cases, the step `2` is just skipped (CI validates commits anyway, so running tests during release does not help).

What I mean is I see no reason in adding `sign` dependency, so I would prefer to avoid adding unrelated dependencies to the plugin.

What is the problem you are trying to solve?

**Comment by vlsi** [comment]
> noticeable long

`org.gradle.parallel=true` solves that, doesn't it?

>  and generating side effects - a new staging repository in Nexus

What's the problem with that?
The repository can easily be dropped.

**Comment by marcphilipp** [comment]
`org.gradle.parallel=true` will only cause tasks from different subprojects to be executed in parallel. Tasks from the same project are always executed sequentially.

**Comment by vlsi** [comment]
>shouldRunAfter(tasks.withType<Sign>()).

Somebody might say: don't bother me asking the pgp password if Nexus repository is not available anyway.

**Comment by marcphilipp** [comment]
I still think preparing the artifacts to be uploaded locally before creating a remote resource is sensible. 🙂 

You could override it with a `mustRunAfter()` on the `Sign` tasks. 😉

**Comment by vlsi** [comment]
> I still think preparing the artifacts to be uploaded locally before creating a remote resource is sensible.

`sign` is not the only task that `prepares artifacts`.
Artifacts can be uploaded in parallel.

If you add `initialize.dependsOn(sign)`, then you basically block parallel upload for multi-project case.

**Comment by marcphilipp** [comment]
You're right, it's not the only artifact that needs to be prepared. Unfortunately, there's no public API to get access to all artifacts and add task dependencies/order constraints for the tasks that build them. Thus, it might be better to just document that this can be done in the build scripts for those users who want this behavior.

**Comment by marcphilipp** [comment]
@szpak WDYT?

**Comment by szpak** [comment]
As we discussed it back then, let's keep it as is and let add - along with way - the point to README explaining how to make signing to be executed first (for people who prefer that way).
